compiler/rustc_codegen_llvm/src/declare.rs

```diff
@@ -141,20 +141,30 @@ impl<'ll, 'tcx> CodegenCx<'ll, 'tcx> {
 
         if self.tcx.sess.is_sanitizer_cfi_enabled() {
             if let Some(instance) = instance {
+                let mut typeids = Vec::new();
                 let typeid = typeid_for_instance(self.tcx, instance, TypeIdOptions::empty());
+                typeids.push(typeid.clone());
                 self.set_type_metadata(llfn, typeid);
                 let typeid =
                     typeid_for_instance(self.tcx, instance, TypeIdOptions::GENERALIZE_POINTERS);
+                typeids.push(typeid.clone());
                 self.add_type_metadata(llfn, typeid);
                 let typeid =
                     typeid_for_instance(self.tcx, instance, TypeIdOptions::NORMALIZE_INTEGERS);
+                typeids.push(typeid.clone());
                 self.add_type_metadata(llfn, typeid);
                 let typeid = typeid_for_instance(
                     self.tcx,
                     instance,
                     TypeIdOptions::GENERALIZE_POINTERS | TypeIdOptions::NORMALIZE_INTEGERS,
                 );
+                typeids.push(typeid.clone());
                 self.add_type_metadata(llfn, typeid);
+                let typeid =
+                    typeid_for_instance(self.tcx, instance, TypeIdOptions::NO_TYPE_ERASURE);
+                if !typeids.contains(&typeid) {
```

Comment:
Why is this one special? If we're only going to emit the unique IDs, that's fine, but why does that rule only apply to the unerased one?

Reply:
If you look at an .ll file when attaching it all, you'll see that the secondary type ids tend to repeat a lot (e.g., they repeat for all non methods). Type id attachments are cheap because the actual type metadata doesn't repeat. However, they make it difficult for us to maintain tests and debug by looking at the .ll files and correlate with the type metadata so I changed it to only add secondary type ids when they're unique. This way it's also easy to identify methods that/when a method have a secondary type id.